### PR TITLE
Migrate VID BDTs to GlobalCache + GBRForest (76X)

### PR DESCRIPTION
**Pull request by lgray**
Forward port of #10556.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

File: RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc [hunk 1]
```diff
@@ -34,50 +34,50 @@ ElectronMVAEstimatorRun2Phys14NonTrig::ElectronMVAEstimatorRun2Phys14NonTrig(con
...
-         << " spp " << _allMVAVars.spp
-         << " etawidth " << _allMVAVars.etawidth
-         << " phiwidth " << _allMVAVars.phiwidth
-         << " OneMinusE1x5E5x5 " << _allMVAVars.OneMinusE1x5E5x5
-         << " R9 " << _allMVAVars.R9
-         << " HoE " << _allMVAVars.HoE
-         << " EoP " << _allMVAVars.EoP
-         << " IoEmIoP " << _allMVAVars.IoEmIoP
-         << " eleEoPout " << _allMVAVars.eleEoPout
+	      << " fbrem "    << vars[11]  //_allMVAVars.fbrem
```
Comment (slava77) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: looking at this somewhat crazy implementation, I'm thinking if it were better to move to TMVAEvaluator ?
How many lines of code and places need to change if one is to insert a new variable.

Comment (lgray) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: TMVA evaluator isn't threadsafe.

-L

(Sent from my Nexus 6)

Comment (slava77) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: Take a look at #10543 .. maybe we can merge to something that fits both?

**Comment by lgray** [comment]
Or, I should say it is threadsafe but you pay for the mutex. Likewise tmva evaluator doesn't yet support gbrforest which is another point of this pr.

**Comment by lgray** [comment]
Ah, didn't see it that pr. Would rather bring that in as an upgrade later.

**Comment by Dr15Jones** [comment]
Don't we also pay a heavy memory tax for TMVA since we switched to ROOT 6?

**Comment by lgray** [comment]
And indeed, changing TMVA evaluatour to be reentrant may require some work. It is pretty stateful right now. 

Would definitely prefer to bring as an update in a later PR. Switching now would degrade memory use and thread efficiency in threaded mode.

File: RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc (hunk 1 above)
Comment (slava77) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: well, my concern here was mostly manual indexing. Maybe define enums instead?
The current row of ```vars[magic number]``` looks pretty bad

**Comment by davidlange6** [comment]
nothing in tmva is different in root6…

File: RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc (hunk 1 above)
Comment (lgray) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: This for a debug statement. I don't use magic numbers in the rest of the
code.

-L

(Sent from my Nexus 6)

Comment (lgray) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: Retract that. OK I'll add some enums later.

-L

(Sent from my Nexus 6)

**Comment by Dr15Jones** [comment]
I thought TMVA used the internal ROOT 'TFormula' which under the hood used CINT in ROOT5 and now cling in ROOT 6. So TMVA itself didn't change but I thought what it depends on did.

File: RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc [hunk 2]
```diff
@@ -213,85 +219,130 @@ void ElectronMVAEstimatorRun2Phys14NonTrig::fillMVAVariables(const edm::Ptr<reco
...
+  } else {
+    vars = std::move( packMVAVariables( allMVAVars.kfhits,
+                                        allMVAVars.see,
+                                        allMVAVars.spp,
+                                        allMVAVars.OneMinusE1x5E5x5,
+                                        allMVAVars.R9,
+                                        allMVAVars.etawidth,
+                                        allMVAVars.phiwidth,
+                                        allMVAVars.HoE,
+                                        allMVAVars.kfchi2,
```
Comment (slava77) [inline RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.cc]: this ~20-line copy-paste with only one line change is yet another poorly maintainable piece.

File: RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Phys14NonTrig.h [hunk 3]
```diff
@@ -62,29 +63,31 @@ class PhotonMVAEstimatorRun2Phys14NonTrig : public AnyMVAEstimatorRun2Base{
...
   // Call this function once after the constructor to declare
   // the needed event content pieces to the framework
-  void setConsumes(edm::ConsumesCollector&&) override;
+  // DEPRECATED
+  void setConsumes(edm::ConsumesCollector&&) const override;
   // Call this function once per event to retrieve all needed
   // event content pices
-  void getEventContent(const edm::Event& iEvent) override;
+  // DEPRECATED
+  // void getEventContent(const edm::Event& iEvent) const override;
```
Comment (slava77) [inline RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Phys14NonTrig.h]: deprecated and commented out means it's not needed anymore and can be removed.
What about other "DEPRECATED" that are still used??

Comment (lgray) [inline RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Phys14NonTrig.h]: It's final in the base class, can't override it anymore.

(Sent from my Nexus 6)

File: RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Spring15NonTrig.h [hunk 4]
```diff
@@ -62,29 +63,29 @@ class PhotonMVAEstimatorRun2Spring15NonTrig : public AnyMVAEstimatorRun2Base{
...
+  int findCategory( const edm::Ptr<reco::Candidate>& particle ) const override;
   // The function below ensures that the variables passed to MVA are 
   // within reasonable bounds
-  void constrainMVAVariables();
+  void constrainMVAVariables(AllVariables&) const;
 
   // Call this function once after the constructor to declare
   // the needed event content pieces to the framework
-  void setConsumes(edm::ConsumesCollector&&) override;
+  void setConsumes(edm::ConsumesCollector&&) const override;
```
Comment (slava77) [inline RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Spring15NonTrig.h]: why no "DEPRECATED" comment here?

File: RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Phys14NonTrig.h (hunk 3 above)
Comment (slava77) [inline RecoEgamma/PhotonIdentification/interface/PhotonMVAEstimatorRun2Phys14NonTrig.h]: remove it here, then?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10618	e87c12e

- code changes for the migration to GBRForest delivers the desired change; 
- there is quite a bit of code replication pairing variables, names, and generic array with the variable values. This is expected to be updated separately, as discussed partly here partly on skype
- jenkins tests complete and comparisons with the baseline show no difference (as intended from this code migration/refactoring)
- local tests to check memory use were done  in CMSSW_7_6_X_2015-08-06-2300 /test area sign577/ on 200 events from 251721 (DoubleMuon-Run2015B-251721 44C77595*.root file):
    - the net reduction in memory (on event 10) is 45 MB:
        - about 7.5 MB in the electron MVAValueMapProducer
        - about 37 MB in the two photon MVAValueMapProducers

**Comment by davidlange6** [comment]
+1

**Pull request merged**